compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/optimizer/BIRBasicBlockOptimizer.java

```diff
@@ -0,0 +1,317 @@
+/*
+ *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.wso2.ballerinalang.compiler.bir.optimizer;
+
+import org.wso2.ballerinalang.compiler.bir.model.BIRNode;
+import org.wso2.ballerinalang.compiler.bir.model.BIRNode.BIRBasicBlock;
+import org.wso2.ballerinalang.compiler.bir.model.BIRTerminator;
+import org.wso2.ballerinalang.compiler.bir.model.BIRVisitor;
+import org.wso2.ballerinalang.compiler.util.Name;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Remove unnecessary goto basic blocks.
+ *
+ * @since 2.0
+ */
+
+public class BIRBasicBlockOptimizer extends BIRVisitor {
+
+    private BIROptimizer.OptimizerEnv env;
+    private final Set<BIRBasicBlock> predecessors = new HashSet<>();
+    private final Map<BIRBasicBlock, List<BIRBasicBlock>> predecessorMap = new HashMap<>();
+
+    private int currentBBId = -1;
+    private static final String BIR_BASIC_BLOCK_PREFIX = "bb";
+
+    public void optimizeNode(BIRNode node, BIROptimizer.OptimizerEnv env) {
+        if (node == null) {
+            return;
```

Comment:
Do we need this null check?

Reply:
There might be a chance to get `null` value for the `node`. Just added this check to omit the risk of null pointer exception and we do not need to create a new environment to keep the previous states for such cases. [But added tests are not covering this line]